src/viam/sdk/services/mlmodel/client.cpp

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 Viam Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <viam/sdk/services/mlmodel/client.hpp>
+
+#include <viam/sdk/services/mlmodel/private/proto.hpp>
+
+#include <grpcpp/channel.h>
+
+namespace viam {
+namespace sdk {
+
+namespace {
+
+struct tensor_storage_and_views {
+    mlmodel_details::tensor_storage storage;
+    MLModelService::named_tensor_views views;
+};
+
+}  // namespace
+
+MLModelServiceClient::MLModelServiceClient(std::string name, std::shared_ptr<grpc::Channel> channel)
+    : MLModelServiceClient(std::move(name), service_type::NewStub(channel)) {
+    channel_ = std::move(channel);
+}
+
+MLModelServiceClient::MLModelServiceClient(std::string name,
+                                           std::unique_ptr<service_type::StubInterface> stub)
+    : MLModelService(name), stub_(std::move(stub)) {}
+
+std::shared_ptr<MLModelService::named_tensor_views> MLModelServiceClient::infer(
+    const named_tensor_views& inputs) {
+    namespace pb = ::google::protobuf;
+    namespace mlpb = ::viam::service::mlmodel::v1;
+
+    pb::Arena arena;
+    auto* const req = pb::Arena::CreateMessage<mlpb::InferRequest>(&arena);
+
+    req->set_name(this->name());
+
+    auto& mutable_input_data = *req->mutable_input_data();
+    auto& mutable_input_data_fields = *mutable_input_data.mutable_fields();
+
+    for (const auto& kv : inputs) {
+        auto ib = mutable_input_data_fields.emplace(kv.first);
+        pb::Value& value = ib.first->second;
+        mlmodel_details::tensor_to_pb_value(kv.second, &value);
+    }
+
+    auto* const resp = pb::Arena::CreateMessage<mlpb::InferResponse>(&arena);
+
+    grpc::ClientContext ctx;
+    auto result = stub_->Infer(&ctx, *req, resp);
+
+    // TODO: Metadata caching.
```

Comment:
This is fairly important because otherwise we need to make a metadata call for each infer call, and we don't want to need to do that. But it wasn't a requirement to have that working in order to move forward to review.

Reply:
Is it safe to cache metadata? What if the model service has been reconfigured?

Reply:
Yes, this is the concern I brought up last week in the meeting. The explanation I received was that it should be OK to cache, because if an mlmodel service is reconfigured then components which depend on it should also be reconfigured, and so the client would be re-created - effectively dropping the cache. If that logic isn't sound, then some time will need to be invested in considering solutions. For now, we fetch the metadata on each inference, which isn't great, and may also not be safe across reconfigurations if those are actually a problem, since you could race between obtaining the metadata and a reconfiguration.

Reply:
This is true for the vision service, when it depends on an ML model service. But this won't be true for an external client using the C++ SDK.

Reply:
That's a very good point which I had not considered. A few ways I see to make it work would be to 1) require fetching metadata on every inference as we are doing now, or 2) move only the necessary subset of metadata into the inference response (e.g. the `shape` info in the proposed flat tensors scope is already a move in this direction), or 3) including a version field in the metadata response and returning the metadata version in the inference response,  which would allow clients to invalidate and refresh the cache. I suspect that 3 is the right answer since 1 is probably unacceptable for performance reasons, and 2 feels vague. The downside to three is that it places requirements on servers to manage the metadata version properly, and on clients to validate and refresh. Definitely interested if you have any other ideas how we might approach the issue.

Meanwhile, I will update the comment to say "Investigate" and fold in some of this discussion.

Reply:
Filed https://viam.atlassian.net/browse/RSDK-3298, and I'll update the comment.